Proposed migration

bryceharrington opened the pull request:
First cut at some initial docs about proposed migration.

The ubuntu wiki has a more authoritative guide, but is structured more as a reference.  This is intending to serve more as a tutorial and a collection point to links to the various sources of information a new +1 maintainer would need.

This still lacks a bit of polish (the glossary section in particular is going to render weirdly), but I think it's ready for an initial review.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
-Non-sru:
+  0.  Issue identified, fixed, and packaged
+  1.  *source.changes uploaded
+  2.  If (and only if) the fix is an SRU:
```
Or if we are in a freeze - should be added plus a link to https://wiki.ubuntu.com/FreezeExceptionProcess ?

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+  0.  Issue identified, fixed, and packaged
+  1.  *source.changes uploaded
+  2.  If (and only if) the fix is an SRU:
+      a.  Upload goes into an 'unapproved' queue
```
These (and the all the latter a/b/c indented lists) do not render correctly in all markdown readers.
It becomes one long convoluted line.
Making it "*" or "1./2./3." would fix that

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+      b.  SRU team reviews and approves... (goto 3)
+      c.  ...or disapproves (goto 0)
+  3.  Upload goes into [codename]-proposed queue
+  4.  Build binary package(s) from source package
```
To avoid misunderstanding maybe "launchpad will build" ?
As e.g. people from Debian are still used to build locally and upload binaries.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+      c.  ...or disapproves (goto 0)
+  3.  Upload goes into [codename]-proposed queue
+  4.  Build binary package(s) from source package
+  5.  Run autopkgtests
```
Add a link to either of
https://packaging.ubuntu.com/html/auto-pkg-test.html
https://salsa.debian.org/ci-team/autopkgtest/blob/master/doc/README.package-tests.rst
https://autopkgtest.ubuntu.com/
here ?

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+      a.  Are binary packages installable?
+      b.  Are all required dependencies at the right version?
+  7.  If (and only if) the fix is an SRU:
+      a.  SRU bug updated with request to verify
```
Add a link to https://wiki.ubuntu.com/StableReleaseUpdates#Verification ?

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
--- all pkg tests are OK
+The build step can hit failures for several general classes of problems.  First are regular build problems that can be easily reproduced locally (and that should have been addressed before uploading).  Second are intermittent platform problems, where simply retrying the build once or twice can result in a successful build.  Third are dependency problems, where the task is to get some other packag ...
+
+The first case covers all the usual normal build issues like syntax errors and so on.  In practice these are typically quite rare because developers tend to be very diligent at making sure their packages build before uploading.  Fixes for these is the usual standard development work.
```
Double space before "Fixes".
You correctly state that those are rare, maybe it is worth to mention that if they happen that is usually on less common platforms - and that one should consider using canonistack to reproduce it?

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+
+The first case covers all the usual normal build issues like syntax errors and so on.  In practice these are typically quite rare because developers tend to be very diligent at making sure their packages build before uploading.  Fixes for these is the usual standard development work.
+
+In the second case, local building was fine but the uploaded build may have failed on Launchpad due to a transitory issue such as a network outage, or a race condition, or other stressed resource.  Or it might have failed due to a strange dependence on some variable like the day of the week.  In these cases, clicking on the "Retry Build" button in Launchpad once or twice can sometimes cause the b ...
```
double space before "Or"

utkarsh2102 replied:
Worth adding that you can't click on "retry" and need a core-dev's assistance, heh.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
+
+In the second case, local building was fine but the uploaded build may have failed on Launchpad due to a transitory issue such as a network outage, or a race condition, or other stressed resource.  Or it might have failed due to a strange dependence on some variable like the day of the week.  In these cases, clicking on the "Retry Build" button in Launchpad once or twice can sometimes cause the b ...
+
+For the third case, where a dependency is not the right version or is not present in the right pocket, the question becomes one of identifying what's wrong with the dependency and fixing it.  Be aware there may be some situations where the problem really is with the dependency itself, and the solution is to change the version of the dependency, or adjust the dependency version requirement, or so  ...
```
double space before "Be".
Hmm - there are a lot of those - maybe they are intentional - in that case keep them.
If not then a simple search and replace will do it I guess.

utkarsh2102 replied:
+1. I'd rather have them removed, not consistent with other pages, or are they?

cpaelzer commented on ProposedMigration.md:
```diff
@@ -1,39 +1,50 @@
-If
+The fourth case of ABI changes comes up particularly when Ubuntu is introducing new versions of language runtime environments or core libraries; i.e. new glibc, ruby, python3, phpunit, et al.  Your package may fail to build because one of its dependencies was built against a different version of glibc or whatever, and needs (no-change) rebuilt (and/or patched) to build with the new version.
 
+Finally, there are also a miscellania of other problems that can result in build failures.  Many of these will be package-specific or situation-specific.  As you run into situations that crop up more than a couple times please update these docs.
```
I think the common-enough-to-mention case here is the changed behavior or toolchain - very similar to your fourth case. Often caused by new glibc, gcc, python3, glib2.0 and such.

This kind essentially tends to happen when the release of the underlying libraries/toolchain is newer than the consuming project. Upstreams need time to adapt to changed behavior or function-prototypes. And in a similar fashion for the compiler toolchain that usually becomes more strict each release. And while one might say "but gcc vX was usable for 2 years; the time upstream projects really realize and care is when Distributions make it the default".
The important hint to add here is that quite often such an update causes the same kind of error in many places.
So if you find a package broken by gcc, glibc or similar update but can reproduce it to be fine with the older releases it is recommended to a) speak up in #ubuntu-devel as it is quite common to find other such cases and all need the same solution that might already exist b) report your findings in a way that others can re-use them if they run into equally affected FTFBS cases.
If upstreams do not respond in time it usually means to file a bug there (for long term resolution) and temporarily ignore the new warning via `-W...` or switch to the old behavior with `-f...`

cpaelzer commented on ProposedMigration.md:
```diff
@@ -17,12 +17,28 @@ Here is the typical lifecycle for an upload:
+
+Many of the items on the page are not actually broken, they're just blocked by something else that needs resolution.  This often happens for new libraries or language runtimes, or really any package that a lot of other things depend on to build, install, or run tests against.
+
+If you are just coming up to speed on proposed migration duty, you'll likely be wondering "Where do I even start?!"  A good suggestion is to look for build failures towards the top of the page; build failures tend to be more localized and more deterministic as to cause, and items towards the top of the page will have had fewer eyeballs on them since they're newer, and so have a higher chance of b ...
```
OTOH it often means someone could have just uploaded (or watch this upload) this and already have an look at it. I'd recommend to at least pinging #ubuntu-devel and the uploader himself when picking from the top.

cpaelzer replied:
Another orthogonal but equally helpful hint I have for "where do I even start" is to ignore the top (too much churn and people look at it) and the bottom (often a lost cause) and instead look at the middle of the list - and there in particular at FTBFS on just one of the architectures. Those tend to be fine code that just didn't think of a particular behavior on other platforms. Those are usually easily reproducible and debuggable in canonistack. Once analyzed upstreams are often happy for the help as they are just not aware of some of those architectures.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -41,6 +41,12 @@ Many of the items on the page are not actually broken, they're just blocked by s
+Bug Reports for Migration Problems
+----------------------------------
+
+If you file a bug report about a build or test failure shown on the update excuses page, tag the bug report 'update-excuse'.  This helps other developers see the investigation work you've already done, and can be used to identify the next-action.  Mark yourself as the Assignee if you're actively working on the issue, and leave the bug unsubscribed if you aren't so that others can carry things for ...
```
Using `update-excuse` instead of 'update-excuse' would help to be more aware of this special tag name.

basak replied:
It might be worth pointing out that doing this results in the bug appearing in the relevant place on the excuses page.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -103,3 +103,111 @@ File a MP against it with a description indicating the lp bug#, rationale for wh
+  approval, b) cannot determine if permanent, or c) permanent rejection.
+
+
+* Has no binaries on any arch (- to x.y.z)
```
I've found that "Has no binaries on arch <foo>" are great candidates for debugging and resolving issues.
I've already mentioned this in detail at your section about "where to start" at processing update-excuses, but a short mention of this type here would be nice as well.

cpaelzer commented on ProposedMigration.md:
```diff
@@ -128,7 +128,24 @@ rm: cannot remove '/var/cache/munin/www/localdomain/localhost.localdomain': Dire
+
+The package's debian/tests/control file defines what gets installed in the test environment before executing the tests.  You can review and verify the packages and versions in the DEP8 test log, between the lines 'autopkgtest...: test integration: preparing testbed' and 'Removing autopkgtest-satdep'.
+
+A common issue is that the test should be run against a version of a dependency present in the -proposed pocket, however it failed due to running against the version in -release.  Often this is straightforward to prove by running the autopkgtests locally in a container.  Another easy way to test this is to re-run the test but set it to preferentially pull packages from -proposed -- this is done b ...
```
Add a link https://bazaar.launchpad.net/~ubuntu-archive/ubuntu-archive-tools/trunk/view/head:/retry-autopkgtest-regressions for retry-autopkgtest-regressions as well?

utkarsh2102 replied:
Also worth mentioning again that a non-core-dev cannot retry anything whatsoever. They'd need assistance of a core-dev.

cpaelzer reviewed (changes requested):
Thanks for this writeup!
I've a few requests for changes/extension here and there, but I'm already +0.8 on this :-)
Have a look what you think about my suggestions and let me know once they are integrated (or explain my why they are not needed).

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+Failure to Build From Source (FTBFS)
+------------------------------------
+
+The build step can hit failures for several general classes of problems.  First are regular build problems that can be easily reproduced locally (and that should have been addressed before uploading).  Second are intermittent platform problems, where simply retrying the build once or twice can result in a successful build.  Third are dependency problems, where the task is to get some other packag ...
```
I'd also mention the architecture-specific problems. For example, ruby2.7 built on all arches except for arm*. So things like that.

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+
+For the third case, where a dependency is not the right version or is not present in the right pocket, the question becomes one of identifying what's wrong with the dependency and fixing it.  Be aware there may be some situations where the problem really is with the dependency itself, and the solution is to change the version of the dependency, or adjust the dependency version requirement, or so  ...
+
+The fourth case of ABI changes comes up particularly when Ubuntu is introducing new versions of language runtime environments or core libraries; i.e. new glibc, ruby, python3, phpunit, et al.  Your package may fail to build because one of its dependencies was built against a different version of glibc or whatever, and needs (no-change) rebuilt (and/or patched) to build with the new version.
```
aw, you didn't mention PHP itself? 😞 
(lol, I am just kidding!)

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+The fourth case of ABI changes comes up particularly when Ubuntu is introducing new versions of language runtime environments or core libraries; i.e. new glibc, ruby, python3, phpunit, et al.  Your package may fail to build because one of its dependencies was built against a different version of glibc or whatever, and needs (no-change) rebuilt (and/or patched) to build with the new version.
+
+Finally, there are also a miscellania of other problems that can result in build failures.  Many of these will be package-specific or situation-specific.  As you run into situations that crop up more than a couple times please update these docs.
+
```
I also would like to mention that even when there seems no apparent reason for a package to migrate in the update_excuses page, the answer might lie in the update_output page, which talks about the uninstallability issues. Having gone through this very recently (and multiple times) and having all the documentation at hand, I think I can write a patch after this one has been merged. Thoughts?

utkarsh2102 commented on ProposedMigration.md:
````diff
@@ -0,0 +1,271 @@
+
+### Other Common Issues ###
+
+Autopkgtest runs tests in a controlled network environment, so if a test case expects to download material from the internet, it will likely fail.  If the test case is attempting to download a dependency (e.g. via PIP or Maven), sometimes this can be worked around by adding the missing dependency to ```debian/tests/control```.  If it is attempting to download an example file, then it may be possi ...
````
Trivial but also worth adding that a bunch of tests fail because of `HTTP_PROXY` issues. I don't know if it's a good fit for here but it has bitten me a lot many times :sigh:

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+Skipping tests
+--------------
+
+If an autopkgtest is badly written, it may be too challenging to get it to pass.  In these extreme cases, its possib\
```
I am afraid that `possib\` isn't the right syntax. Please make it "possible" again and shift to the next line, markdown will automatically understand what you wanted to do.

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+
+Checkout [https://git.launchpad.net/~ubuntu-release/britney/+git/hints-ubuntu](lp:~ubuntu-release/britney/hints-ubuntu)
+
+File a MP against it with a description indicating the lp bug#, rationale for why the test can and should be skipped\
```
Redundant "\" at the end, please remove this? 😅

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+File a MP against it with a description indicating the lp bug#, rationale for why the test can and should be skipped\
+, and explanation of what will be unblocked to migration.
+
+Reviewers should be 'canonical-server', 'ubuntu-release', and any archive admins or foundations team members you've \
```
Redundant "\" at the end, please remove this? 😅

utkarsh2102 commented on ProposedMigration.md:
```diff
@@ -0,0 +1,271 @@
+If an autopkgtest is badly written, it may be too challenging to get it to pass.  In these extreme cases, its possib\
+le to request that test failures be ignored for purposes of package migration.
+
+Checkout [https://git.launchpad.net/~ubuntu-release/britney/+git/hints-ubuntu](lp:~ubuntu-release/britney/hints-ubuntu)
```
Missing fullstop in the end? :P

utkarsh2102 reviewed (changes requested):
Awesome stuff, Bryce! 🎉 

Few concerns and suggestions inline. Besides, two more generic points:
- can you please remove the double spaces b/w the sentences?
- can you also wrap up lines in 80 words or so? Maybe you can skip this here for now, I'll format the entire handbook to do that for easy readability, I believe. Up to you.

Thank you for working on this!

bryceharrington commented:
Thanks for the feedback Christian!  Esp. all the hyperlinks, which will
help make this doc a good jump off point to all the various information
sites.

On Tue, Jul 06, 2021 at 11:46:01PM -0700, Christian Ehrhardt wrote:
> @cpaelzer requested changes on this pull request.
> 
> Thanks for this writeup!
> I've a few requests for changes/extension here and there, but I'm already +0.8 on this :-)
> Have a look what you think about my suggestions and let me know once they are integrated (or explain my why they are not needed).
> 
> >  
> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> 
> Or if we are in a freeze - should be added plus a link to https://wiki.ubuntu.com/FreezeExceptionProcess ?

Incorporated

> >  
> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> +      a.  Upload goes into an 'unapproved' queue
> 
> These (and the all the latter a/b/c indented lists) do not render correctly in all markdown readers.
> It becomes one long convoluted line.
> Making it "*" or "1./2./3." would fix that

That's too bad.  Guess I'll change the sub-level to *'s.

> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> +      a.  Upload goes into an 'unapproved' queue
> +      b.  SRU team reviews and approves... (goto 3)
> +      c.  ...or disapproves (goto 0)
> +  3.  Upload goes into [codename]-proposed queue
> +  4.  Build binary package(s) from source package
> 
> To avoid misunderstanding maybe "launchpad will build" ?
> As e.g. people from Debian are still used to build locally and upload binaries.

Incorporated

> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> +      a.  Upload goes into an 'unapproved' queue
> +      b.  SRU team reviews and approves... (goto 3)
> +      c.  ...or disapproves (goto 0)
> +  3.  Upload goes into [codename]-proposed queue
> +  4.  Build binary package(s) from source package
> +  5.  Run autopkgtests
> 
> Add a link to either of
> https://packaging.ubuntu.com/html/auto-pkg-test.html
> https://salsa.debian.org/ci-team/autopkgtest/blob/master/doc/README.package-tests.rst
> https://autopkgtest.ubuntu.com/
> here ?

I'll add all three links, but at this particular spot in the doc the
first link feels most appropriate.  I'll work the other two in below.

> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> +      a.  Upload goes into an 'unapproved' queue
> +      b.  SRU team reviews and approves... (goto 3)
> +      c.  ...or disapproves (goto 0)
> +  3.  Upload goes into [codename]-proposed queue
> +  4.  Build binary package(s) from source package
> +  5.  Run autopkgtests
> +  6.  Verify other archive consistency checks
> +      a.  Are binary packages installable?
> +      b.  Are all required dependencies at the right version?
> +  7.  If (and only if) the fix is an SRU:
> +      a.  SRU bug updated with request to verify
> 
> Add a link to https://wiki.ubuntu.com/StableReleaseUpdates#Verification ?

Incorporated

> -When tests succeeded, built fine:
> -- In disco, it would migrate from proposed to release
> -- special cases: not installable, waiting on dependency
> -3 things needed:
> --- not installable: depends on something not existing, for example
> --- dependency: Dependent package has not gone into release yet.
> --- all pkg tests are OK
> +The build step can hit failures for several general classes of problems.  First are regular build problems that can be easily reproduced locally (and that should have been addressed before uploading).  Second are intermittent platform problems, where simply retrying the build once or twice can result in a successful build.  Third are dependency problems, where the task is to get some other package to be the right version in the release pocket so that the build will succeed.  Fourth are ABI changes with dependencies; these often just need a no-change rebuild.  And last are all the other myriad package-specific build problems, that typically require some particular tweak to the package itself.
> +
> +The first case covers all the usual normal build issues like syntax errors and so on.  In practice these are typically quite rare because developers tend to be very diligent at making sure their packages build before uploading.  Fixes for these is the usual standard development work.
> 
> Double space before "Fixes".

I've always been a doublespace-between-sentences guy and probably always
will.  But I believe Markdown automatically translates them into
appropriate spacing during formatting, so probably not a worry.

> You correctly state that those are rare, maybe it is worth to mention that if they happen that is usually on less common platforms - and that one should consider using canonistack to reproduce it?

Sure, incorporated:

"The second case is similar to the first but pertains to issues that arise only on specific platforms.  Common situations are tests or code that relies on [https://en.wikipedia.org/wiki/Endianness](endianness) of data types and so breaks on big-endian systems (e.g. s390x), code expecting 64-bit data types that breaks on 32-bit (e.g. armhf), and so on.  [https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01](Canonistack provides hardware) that Canonical employees can use to attempt reproduction of these problems.  Take note with i386 particularly in that it often fails less due to data type issues but more because it is a partial port and has numerous limitations that require special handling, as described on the [https://wiki.ubuntu.com/i386](i386 troubleshooting page)."

I'm not sure about linking to intranet docs from a doc that will be
externally visible, but it seems not worth mentioning Canonistack
without the corresponding landing page for how to use it.

I also took this opportunity to tack on some info about i386 and how
its platform issues are bit different than just architectural, with a
link to the right page for those issues.  i386 might be worth a full
paragraph itself, but maybe just this sentence is enough.

> +For the third case, where a dependency is not the right version or is not present in the right pocket, the question becomes one of identifying what's wrong with the dependency and fixing it.  Be aware there may be some situations where the problem really is with the dependency itself, and the solution is to change the version of the dependency, or adjust the dependency version requirement, or so forth.  These latter solutions often require asking for an archive admin's help on the #ubuntu-release IRC channel.
> 
> double space before "Be".
> Hmm - there are a lot of those - maybe they are intentional - in that case keep them.
> If not then a simple search and replace will do it I guess.

They're intentional, but if someone cleaned them up later it wouldn't
bother me.  It's too much muscle memory for me now.

> +Finally, there are also a miscellania of other problems that can result in build failures.  Many of these will be package-specific or situation-specific.  As you run into situations that crop up more than a couple times please update these docs.
> 
> I think the common-enough-to-mention case here is the changed behavior
> or toolchain - very similar to your fourth case. Often caused by new
> glibc, gcc, python3, glib2.0 and such.

This is what I intend the fifth case to be, so will fold your notes into
that one.  From a +1 maintenance POV a toolchain/api change is a much
more sigificant problem than just an incorrect dependency version.  Both
are quite common problems, but need distinct workflows to address them.

> This kind essentially tends to happen when the release of the underlying libraries/toolchain is newer than the consuming project. Upstreams need time to adapt to changed behavior or function-prototypes. And in a similar fashion for the compiler toolchain that usually becomes more strict each release. And while one might say "but gcc vX was usable for 2 years; the time upstream projects really realize and care is when Distributions make it the default".
> The important hint to add here is that quite often such an update causes the same kind of error in many places.
> So if you find a package broken by gcc, glibc or similar update but can reproduce it to be fine with the older releases it is recommended to a) speak up in #ubuntu-devel as it is quite common to find other such cases and all need the same solution that might already exist b) report your findings in a way that others can re-use them if they run into equally affected FTFBS cases.
> If upstreams do not respond in time it usually means to file a bug there (for long term resolution) and temporarily ignore the new warning via `-W...` or switch to the old behavior with `-f...`

Cool thanks, this adds a ton of info for this fifth case.  Since this
one is challenging to deal with yet very common, this is probably going
to be very useful material.  I've rephrased and rewritten the above to
fit it in, so will surely be worth re-review.


> +If you are just coming up to speed on proposed migration duty, you'll likely be wondering "Where do I even start?!"  A good suggestion is to look for build failures towards the top of the page; build failures tend to be more localized and more deterministic as to cause, and items towards the top of the page will have had fewer eyeballs on them since they're newer, and so have a higher chance of being something simple.
> 
> OTOH it often means someone could have just uploaded (or watch this upload) this and already have an look at it. I'd recommend to at least pinging #ubuntu-devel and the uploader himself when picking from the top.
> 

Ah yes, I get out of the habit of asking, since so many go inactive
during my timezone, and forget that strategy probably works better for
others.  I'll add it.

> +If you are just coming up to speed on proposed migration duty, you'll likely be wondering "Where do I even start?!"  A good suggestion is to look for build failures towards the top of the page; build failures tend to be more localized and more deterministic as to cause, and items towards the top of the page will have had fewer eyeballs on them since they're newer, and so have a higher chance of being something simple.
> 
> Another orthogonal but equally helpful hint I have for "where do I even start" is to ignore the top (too much churn and people look at it) and the bottom (often a lost cause) and instead look at the middle of the list - and there in particular at FTBFS on just one of the architectures. Those tend to be fine code that just didn't think of a particular behavior on other platforms. Those are usually easily reproducible and debuggable in canonistack. Once analyzed upstreams are often happy for the help as they are just not aware of some of those architectures.
> 

Incorporated


> +Bug Reports for Migration Problems
> +----------------------------------
> +
> +If you file a bug report about a build or test failure shown on the update excuses page, tag the bug report 'update-excuse'.  This helps other developers see the investigation work you've already done, and can be used to identify the next-action.  Mark yourself as the Assignee if you're actively working on the issue, and leave the bug unsubscribed if you aren't so that others can carry things forward.
> 
> Using `update-excuse` instead of 'update-excuse' would help to be more aware of this special tag name.

Thanks.  I bet there's a lot of places throughout the handbook that
could benefit from this formatting.

> +* Has no binaries on any arch (- to x.y.z)
> 
> I've found that "Has no binaries on arch <foo>" are great candidates for debugging and resolving issues.
> I've already mentioned this in detail at your section about "where to start" at processing update-excuses, but a short mention of this type here would be nice as well.

Incorporated

> +A common issue is that the test should be run against a version of a dependency present in the -proposed pocket, however it failed due to running against the version in -release.  Often this is straightforward to prove by running the autopkgtests locally in a container.  Another easy way to test this is to re-run the test but set it to preferentially pull packages from -proposed -- this is done by appending '&all-proposed=1' to the test URL.  If that passes, but the package still does not migrate, then look in the test log for all packages that were pulled from -proposed and include those as triggers.  [https://git.launchpad.net/~bryce/+git/excuses-kicker](Excuses Kicker) and retry-autopkgtest-regressions are handy tools for generating these URLs.
> 
> Add a link https://bazaar.launchpad.net/~ubuntu-archive/ubuntu-archive-tools/trunk/view/head:/retry-autopkgtest-regressions for retry-autopkgtest-regressions as well?
> 

Thanks, incorporated

Bryce

bryceharrington commented:
On Wed, Jul 07, 2021 at 12:43:21AM -0700, Utkarsh Gupta wrote:
> @utkarsh2102 requested changes on this pull request.
> 
> Awesome stuff, Bryce! 🎉 
> 
> Few concerns and suggestions inline. Besides, two more generic points:
> - can you please remove the double spaces b/w the sentences?
> - can you also wrap up lines in 80 words or so? Maybe you can skip this here for now, I'll format the entire handbook to do that for easy readability, I believe. Up to you.
> 

Heh, I'll leave that formatting to you.

I wasn't sure if markdown would properly format line-wrapped text and
have been too lazy to test it.  If it works, then I'd definitely +1 that
change.

> +Often the migration proceeds automatically, but when there are issues our involvement is needed to sort things out.  Sometimes the package having the problem is one we've uploaded ourselves, so naturally want to work on solving the issue.  But in other times the issue arises organically such as via something that auto-sync'd from Debian, or a side-effect from some other change in the distro, and thus have no defined "owner".  For these latter problems (and for the former when they've become tricky), the Ubuntu project requires distro developers to devote a portion of their time to focusing on migration issues generally, on a rotating basis each working a few days or week at a time.
> 
> Maybe add "called +1" or something similar at last?

I didn't want to have to get into why it's called "+1 maintenance" so
intentionally just kept to a generic term for it.

> Also, aren't these just for Canonical employees and not Ubuntu devs in general?

This document is geared to Canonical employees so yeah that's going to
be 90% of the audience, however the document is public so does provide
reference for the community generally (IOW there's no reason a community
person *couldn't* use this information).

Also, while +1 maintainers would be the primary consumer of this
information, the intent is to also provide a reference for folks working
on transitions or just trying to figure out WTF their fix is stuck in
-proposed.

> +Failure to Build From Source (FTBFS)
> +------------------------------------
> +
> +The build step can hit failures for several general classes of problems.  First are regular build problems that can be easily reproduced locally (and that should have been addressed before uploading).  Second are intermittent platform problems, where simply retrying the build once or twice can result in a successful build.  Third are dependency problems, where the task is to get some other package to be the right version in the release pocket so that the build will succeed.  Fourth are ABI changes with dependencies; these often just need a no-change rebuild.  And last are all the other myriad package-specific build problems, that typically require some particular tweak to the package itself.
> 
> I'd also mention the architecture-specific problems. For example, ruby2.7 built on all arches except for arm*. So things like that.

Incorporated

> +The fourth case of ABI changes comes up particularly when Ubuntu is introducing new versions of language runtime environments or core libraries; i.e. new glibc, ruby, python3, phpunit, et al.  Your package may fail to build because one of its dependencies was built against a different version of glibc or whatever, and needs (no-change) rebuilt (and/or patched) to build with the new version.
> 
> aw, you didn't mention PHP itself? 😞 
> (lol, I am just kidding!)

Hah I can't always resist but did this time.  :-)

> +Finally, there are also a miscellania of other problems that can result in build failures.  Many of these will be package-specific or situation-specific.  As you run into situations that crop up more than a couple times please update these docs.
> +
> 
> I also would like to mention that even when there seems no apparent reason for a package to migrate in the update_excuses page, the answer might lie in the update_output page, which talks about the uninstallability issues. Having gone through this very recently (and multiple times) and having all the documentation at hand, I think I can write a patch after this one has been merged. Thoughts? 

Yes, thanks that would be great.

I opted to ignore it since the use cases needing it are (IMHO) rare, but
you're right that they do come up time to time.  The existing docs
(https://wiki.ubuntu.com/ProposedMigration#The_update_output.txt_file_is_completely_unreadable.21)
are not very thorough, unfortunately.  Perhaps a good angle would be to
explain the specific use cases in paint-by-number fashion so you don't
have to document how to actually decode update_output.txt.


> +### Other Common Issues ###
> +
> +Autopkgtest runs tests in a controlled network environment, so if a test case expects to download material from the internet, it will likely fail.  If the test case is attempting to download a dependency (e.g. via PIP or Maven), sometimes this can be worked around by adding the missing dependency to ```debian/tests/control```.  If it is attempting to download an example file, then it may be possible to make the test case use a local file, or to load from the proxy network.
> 
> Trivial but also worth adding that a bunch of tests fail because of `HTTP_PROXY` issues. I don't know if it's a good fit for here but it has bitten me a lot many times :sigh:

Agreed they do come up now and then, and are perfect types of items to
document here.  I alluded to proxy issues vaguely, but perhaps you could
elaborate with some tangible advice (and/or examples) for dealing with
it?  If there's enough to say on it, may be worth calling out with its
own header above this one.

> +Skipping tests
> +--------------
> +
> +If an autopkgtest is badly written, it may be too challenging to get it to pass.  In these extreme cases, its possib\
> 
> I am afraid that `possib\` isn't the right syntax. Please make it "possible" again and shift to the next line, markdown will automatically understand what you wanted to do.
> 

Ah thanks, yes the stray line breaks must have come from some cut and
pasting in emacs.

> +Skipping tests
> +--------------
> +
> +If an autopkgtest is badly written, it may be too challenging to get it to pass.  In these extreme cases, its possib\
> +le to request that test failures be ignored for purposes of package migration.
> +
> +Checkout [https://git.launchpad.net/~ubuntu-release/britney/+git/hints-ubuntu](lp:~ubuntu-release/britney/hints-ubuntu)
> 
> Missing fullstop in the end? :P

I'll rewrite this section.  This text is verbatim from Karl's
misc. notes, but should be revised to discuss hinting more generally.
I'll do that as a new commit in this thread.

> -Non-sru:
> +  0.  Issue identified, fixed, and packaged
> +  1.  *source.changes uploaded
> +  2.  If (and only if) the fix is an SRU:
> +      a.  Upload goes into an 'unapproved' queue
> +      b.  SRU team reviews and approves... (goto 3)
> +      c.  ...or disapproves (goto 0)
> +  3.  Upload goes into [codename]-proposed queue
> +  4.  Build binary package(s) from source package
> +  5.  Run autopkgtests
> 
> Yeah, besides the autopkgtests of the package itself, the autopkgtests of its reverse{-build}-dependencies are run, too.
> 

Incorporated

> +In the second case, local building was fine but the uploaded build may have failed on Launchpad due to a transitory issue such as a network outage, or a race condition, or other stressed resource.  Or it might have failed due to a strange dependence on some variable like the day of the week.  In these cases, clicking on the "Retry Build" button in Launchpad once or twice can sometimes cause the build to randomly succeed, thus allowing the transition to proceed.  Needless to say, we don't wish for such unpredictabilities in the build process, so if the problem reoccurs and if you can narrow down the cause, it can be worthwhile to find the root cause and figure out how to fix it properly.
> 
> Worth adding that you can't click on "retry" and need a core-dev's assistance, heh.

Incorporated

> +For the third case, where a dependency is not the right version or is not present in the right pocket, the question becomes one of identifying what's wrong with the dependency and fixing it.  Be aware there may be some situations where the problem really is with the dependency itself, and the solution is to change the version of the dependency, or adjust the dependency version requirement, or so forth.  These latter solutions often require asking for an archive admin's help on the #ubuntu-release IRC channel.
> 
> +1. I'd rather have them removed, not consistent with other pages, or are they?

Sorry, not sure I understand what you mean here.  Do you mean remove
binaries?  Yeah that would be good to mention here.

> +A common issue is that the test should be run against a version of a dependency present in the -proposed pocket, however it failed due to running against the version in -release.  Often this is straightforward to prove by running the autopkgtests locally in a container.  Another easy way to test this is to re-run the test but set it to preferentially pull packages from -proposed -- this is done by appending '&all-proposed=1' to the test URL.  If that passes, but the package still does not migrate, then look in the test log for all packages that were pulled from -proposed and include those as triggers.  [https://git.launchpad.net/~bryce/+git/excuses-kicker](Excuses Kicker) and retry-autopkgtest-regressions are handy tools for generating these URLs.
> 
> Also worth mentioning again that a non-core-dev cannot retry anything whatsoever. They'd need assistance of a core-dev.

Incorporated.

Thanks again,
bryce

bryceharrington commented:
On Wed, Jul 07, 2021 at 03:01:16AM -0700, basak wrote:
> @basak commented on this pull request.
> 
> 
> 
> > @@ -41,6 +41,12 @@ Many of the items on the page are not actually broken, they're just blocked by s
>  If you are just coming up to speed on proposed migration duty, you'll likely be wondering "Where do I even start?!"  A good suggestion is to look for build failures towards the top of the page; build failures tend to be more localized and more deterministic as to cause, and items towards the top of the page will have had fewer eyeballs on them since they're newer, and so have a higher chance of being something simple.
>  
>  
> +Bug Reports for Migration Problems
> +----------------------------------
> +
> +If you file a bug report about a build or test failure shown on the update excuses page, tag the bug report 'update-excuse'.  This helps other developers see the investigation work you've already done, and can be used to identify the next-action.  Mark yourself as the Assignee if you're actively working on the issue, and leave the bug unsubscribed if you aren't so that others can carry things forward.
> 
> It might be worth pointing out that doing this results in the bug appearing in the relevant place on the excuses page.
> 

Ah thanks, yes forgot that important detail.

Bryce

basak commented:
On Wed, Jul 07, 2021 at 12:41:38PM -0700, Bryce Harrington wrote:
> > Also, aren't these just for Canonical employees and not Ubuntu devs in general?
> 
> This document is geared to Canonical employees so yeah that's going to
> be 90% of the audience, however the document is public so does provide
> reference for the community generally (IOW there's no reason a community
> person *couldn't* use this information).
> 
> Also, while +1 maintainers would be the primary consumer of this
> information, the intent is to also provide a reference for folks working
> on transitions or just trying to figure out WTF their fix is stuck in
> -proposed.

It's still misleading to say "the Ubuntu project requires distro
developers" though. That's objectively not true. Even if this is only
read by Canonical employees, this might lead to new Canonical employees
leaning on non-Canonical Ubuntu developers to do work that they wrongly
think they are required to do.

How about "Canonical requires some of its distro developer employees"
instead?

bryceharrington commented:
On Wed, Jul 07, 2021 at 12:56:22PM -0700, basak wrote:
> On Wed, Jul 07, 2021 at 12:41:38PM -0700, Bryce Harrington wrote:
> > > Also, aren't these just for Canonical employees and not Ubuntu devs in general?
> > 
> > This document is geared to Canonical employees so yeah that's going to
> > be 90% of the audience, however the document is public so does provide
> > reference for the community generally (IOW there's no reason a community
> > person *couldn't* use this information).
> > 
> > Also, while +1 maintainers would be the primary consumer of this
> > information, the intent is to also provide a reference for folks working
> > on transitions or just trying to figure out WTF their fix is stuck in
> > -proposed.
> 
> It's still misleading to say "the Ubuntu project requires distro
> developers" though. That's objectively not true. Even if this is only
> read by Canonical employees, this might lead to new Canonical employees
> leaning on non-Canonical Ubuntu developers to do work that they wrongly
> think they are required to do.
> 
> How about "Canonical requires some of its distro developer employees"
> instead?

Sure, incorporated, thanks.

Bryce

cpaelzer commented:
The email replies work fine as a conversation, but we had plenty of open review threads up here.
I've checked and you have indeed addressed them all - thanks!
To me it looks good now, not sure if you want a review of the new revised content before merging? IIRC @sergiodj wanted to have a look as well - now might be a good time.

bryceharrington commented:
On Wed, Jul 07, 2021 at 11:43:29PM -0700, Christian Ehrhardt wrote:
> The email replies work fine as a conversation, but we had plenty of open review threads up here.
> I've checked and you have indeed addressed them all - thanks!
> To me it looks good now, not sure if you want a review of the new revised content before merging? IIRC @sergiodj wanted to have a look as well - now might be a good time.

I think it's good to land now.  Sergio is a bit tied up with stuff so
sounds like he wouldn't be able to get to it soon.  I'm sure we'll have
refinements in the future so he can follow up with his own ideas when
he's ready.

I will need your help landing it though, as I don't think I've got the
permissions to do so in our gitlab.

Bryce

utkarsh2102 commented:
Alrighty, here we go..

bryceharrington commented:
Thanks!

On Thu, Jul 08, 2021 at 08:45:50AM -0700, Utkarsh Gupta wrote:
> Merged #17 into master.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/canonical/ubuntu-maintainers-handbook/pull/17#event-4995806352
